Models/IonAuthModel.php

```diff
@@ -0,0 +1,2798 @@
+		{
+			if ($this->verifyPassword($password, $user->password, $identity))
+			{
+				if ($user->active == 0)
+				{
+					$this->triggerEvents('post_login_unsuccessful');
+					$this->setError('IonAuth.login_unsuccessful_not_active');
+
+					return false;
+				}
+
+				$this->setSession($user);
+
+				$this->updateLastLogin($user->id);
+
+				$this->clearLoginAttempts($identity);
+				$this->clearForgottenPasswordCode($identity);
+
+				if ($this->config->rememberUsers)
+				{
+					if ($remember)
+					{
+						$this->rememberUser($identity);
+					}
+					else
+					{
+						$this->clearRememberCode($identity);
+					}
+				}
+
+				// Rehash if needed
+				$this->rehashPasswordIfNeeded($user->password, $identity, $password);
+
+				// Regenerate the session (for security purpose: to avoid session fixation)
+				$this->session->regenerate(false);
+
+				$this->triggerEvents(['post_login', 'post_login_successful']);
+				$this->setMessage('IonAuth.login_successful');
+
+				return true;
+			}
+		}
+
+		// Hash something anyway, just to take up time
+		$this->hashPassword($password);
+
+		$this->increaseLoginAttempts($identity);
+
+		$this->triggerEvents('post_login_unsuccessful');
+		$this->setError('IonAuth.login_unsuccessful');
+
+		return false;
+	}
+
+	/**
+	 * Verifies if the session should be rechecked according to the configuration item recheckTimer. If it does, then
+	 * it will check if the user is still active
+	 * @return bool
+	 */
+	public function recheckSession()
+	{
+		$recheck = (NULL !== $this->config->recheckTimer) ? $this->config->recheckTimer : 0;
+
+		if ($recheck !== 0)
+		{
+			$last_login = $this->session->get('last_check');
+			if ($last_login + $recheck < time())
+			{
+				$query = $this->db->select('id')
+								  ->where([
+									  $this->identity_column => $this->session->get('identity'),
+									  'active' => '1'
+								  ])
+								  ->limit(1)
+								  ->orderBy('id', 'desc')
+								  ->get($this->tables['users']);
+				if ($query->numRows() === 1)
+				{
+					$this->session->set_userdata('last_check', time());
+				}
+				else
+				{
+					$this->triggerEvents('logout');
+
+					$identity = $this->config->identity;
+
+					$this->session->unset_userdata([$identity, 'id', 'user_id']);
+
+					return false;
+				}
+			}
+		}
+
+		return (bool)session('identity');
+	}
+
+	/**
+	 * Check if max login attempts exceeded
+	 * Based on code from Tank Auth, by Ilya Konyukhov (https://github.com/ilkon/Tank-Auth)
+	 *
+	 * @param string      $identity  user's identity
+	 * @param string|null $ipAddress IP address
+	 *                               Only used if trackLoginIpAddress is set to TRUE.
+	 *                               If NULL (default value), the current IP address is used.
+	 *                               Use getLastAttemptIp($identity) to retrieve a user's last IP
+	 *
+	 * @return boolean
+	 */
+	public function isMaxLoginAttemptsExceeded(string $identity, string $ipAddress = null): bool
```

Comment:
Remove typehint due to mixed
```suggestion
	public function isMaxLoginAttemptsExceeded(string $identity, $ipAddress = null): bool
```